src/file_utils.R

```diff
@@ -170,6 +182,52 @@ zip_meteo_groups <- function(outfile, xwalk_meteo_fl_names, grouped_meteo_fls){
   scipiper::sc_indicate(outfile, data_file = data_files)
 }
 
+filter_resample_obs <- function(outfile, obs_feather, site_ids, obs_start, obs_stop, sample_res = 0.5){
+  
+  feather::read_feather(obs_feather) %>%
+    filter(site_id %in% site_ids) %>%
+    filter(date >= obs_start & date <= obs_stop) %>%
+    group_by(date, depth, site_id, source) %>%
+    summarize(temp = mean(temp)) %>%
+    ungroup() %>%
+    # interpolate
+    group_by(site_id, date, source) %>%
+    do({
+      date_df <- .
+      date_df %>%
+        mutate(
+          # find the closest PGDL depth to the observation depth
+          new_depth = purrr::map_dbl(depth, function(obsdep) {
+            # get depths from 0 to max depth of obs, rounded up to nearest sample_res, but not beyond
+            # so if obsdep = c(0.3, 9.6), we'll go to 10m, but if obsdep = c(0.3, 9.5), we'd go to 9.5
+            depths <- seq(0, max(ceiling(obsdep / sample_res) * sample_res), by = sample_res)
+            depths[which.min(abs(obsdep - depths))]
+            }),
+          # estimate temperature at the new depth using interpolation, or if
+          # that's not possible, set to the nearest observed temperature
+          new_temp = if(nrow(date_df) >= 2) {
+            # if we have enough values on this date, interpolate
+            approx(x=depth, y=temp, xout=new_depth, rule=2)$y
+          } else {
+            # if we can't interpolate, just use the nearest value
+            temp
+          },
+          depth_diff = abs(depth - new_depth)) %>%
+        # after approx(), trash any values at new_depth >= 0.5 m from the nearest observation
+        filter(depth_diff < 0.5) %>%
+        # only keep one estimate for each new_depth
+        group_by(new_depth) %>%
+        filter(depth_diff == min(depth_diff)) %>%
+        ungroup()
+    }) %>%
+    ungroup() %>%
+    # to see my work as columns, print out the result up to this point, e.g. by uncommenting
+    # tail(20) %>% print(n=20)
+    # now we clean up the columns
+    select(site_id, date, depth=new_depth, temp=new_temp, source) %>%
```

Comment:
@aappling-usgs how about adding this here to deal w/ the duplication issue?

```r
group_by(site_id, date, depth) %>%
    summarize(temp = first(temp), source = first(source)) %>%
    ungroup() %>% 
    saveRDS(file = outfile)
```

Reply:
Good call - I forgot we were grouping by source above, so this deduplication does indeed need to happen outside of that `do()`. Does it make sense to prefer a specific source, e.g., the one with the most observations, instead of just picking the first one?

Reply:
:) it may in the future, but I think this is such an edge-case that I don't want to try to code it in for this experiment. Also, some high-obs sources are worse in quality than ones with fewer...so it may not return the "best" data afterall (?) As long as I am tracking the source used all the way into the model_matched_obs data (which we are), I am ok making this simplification.

Reply:
Sounds totally reasonable.